server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,309 @@
+         */
+        try {
+            this.watchService = PathUtils.getDefaultFileSystem().newWatchService();
+            if (Files.exists(settingsDir)) {
+                Path settingsFilePath = operatorSettingsFile();
+                if (Files.exists(settingsFilePath)) {
+                    logger.info("found initial operator settings file [{}], applying...", settingsFilePath);
+                    // we make a distinction here for startup, so that if we had operator settings before the node started
+                    // we would fail startup.
+                    processFileSettings(settingsFilePath, onStartup);
+                }
+                enableSettingsWatcher(settingsDir);
+            } else {
+                logger.info("operator settings directory [{}] not found, will watch for its creation...", settingsDir);
+                enableSettingsWatcher(environment.configFile());
+            }
+        } catch (Exception e) {
+            if (watchService != null) {
+                try {
+                    this.watchService.close();
+                } catch (Exception ignore) {} finally {
+                    this.watchService = null;
+                }
+            }
+
+            throw new IllegalStateException("unable to launch a new watch service", e);
+        }
+
+        this.watcherThreadLatch = new CountDownLatch(1);
+
+        new Thread(() -> {
+            try {
+                logger.info("file settings service up and running [tid={}]", Thread.currentThread().getId());
+
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    /**
+                     * Reading and interpreting watch service events can vary from platform to platform. E.g:
+                     * MacOS symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_MODIFY:operator
+                     *     ENTRY_CREATE:settings.json
+                     *     ENTRY_MODIFY:settings.json
+                     * Linux in Docker symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_CREATE:operator
+                     * After we get an indication that something has changed, we check the timestamp, file id,
+                     * real path of our desired file.
+                     */
+                    if (Files.exists(settingsDir)) {
+                        try {
+                            Path path = operatorSettingsFile();
+
+                            if (logger.isDebugEnabled()) {
+                                key.pollEvents().stream().forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
+                            }
+
+                            key.pollEvents();
+                            key.reset();
+
+                            enableSettingsWatcher(settingsDir);
```

Comment:
Is there any cost or potential negative behavior from calling register after every take()? ie, why not only do this when the key indicates the parent dir was created?

Reply:
I'll double check to see what happens when the watcher is over and over registered on the same dir, I didn't see an issue when testing, but it doesn't mean it doesn't leak somehow. 

The reason I enable the settings watcher every time is because of symlinks. It's platform specific whether the watcher continues to watch the original symlinked directory or the new one after a change to a link. I couldn't think of another way to make this more reliable, other than repeat that I want to watch the same dir.